Fixes #10782 - global host status

ares opened the pull request:
Keeping my commits separately for easier reviewing. There are two things I didn't refactor in this PR, if we agree it's a good direction I'll create redmine issues.

app/models/report.rb line 66
it would be very nice to refactor this behaviour and ensure the status setter always gets integer number from the importer

app/services/report_status_calculator.rb line 56
replace dynamic define_method and use something like status_of('applied') everywhere instead (I don't think this is highly important)

domcleal commented:
Could you fix the commit messages, so as not to fight the system?  Thanks.  I'll also reset the review label as I don't believe it's been reviewed yet.

ares commented:
I fixed messages. I removed `Not yet reviewed` since you did the same previously so I thought you might have seen it already and hadn't found anything ( :laughing: )

domcleal commented:
Hah.  No, probably a misclick.

domcleal commented on db/migrate/20150612135546_create_host_status.rb:
```diff
@@ -0,0 +1,32 @@
+class CreateHostStatus < ActiveRecord::Migration
+  def self.up
```
No "self." on these methods any more.

domcleal commented on app/views/api/v2/hosts/main.json.rabl:
```diff
@@ -9,7 +9,7 @@ attributes :ip, :environment_id, :environment_name, :last_report, :mac, :realm_i
            :compute_profile_id, :compute_profile_name, :capabilities, :provision_method,
            :puppet_proxy_id, :certname, :image_id, :image_name, :created_at, :updated_at,
-           :last_compile, :puppet_status
+           :last_compile, :global_status
```
Can you keep both?  Alias puppet_status to global_status, and add global_status too.

You'll probably want to add some child nodes for the more detailed statuses.

ares replied:
Would it make sense to display all statuses even if they are not relevant, so keys in output are consistent for all hosts? Or should I rather wrap them in conditions?

domcleal replied:
It makes slightly more sense to me to leave them out if not applicable, but I don't think it makes a lot of difference

domcleal commented on app/services/report_status_calculator.rb:
```diff
@@ -17,20 +18,48 @@ def calculate
+  end
+
+  def status_of(counter)
+    raise(Foreman::Exception.new(N_("invalid type %s") % counter)) unless Report::METRIC.include?(counter)
```
pass `counter` as an argument, don't use %

domcleal commented on app/models/host_status/configuration_status.rb:
```diff
@@ -0,0 +1,85 @@
+    delegate(*Report::METRIC, :to => :calculator)
+
+    def out_of_sync?
+      if (host && !host.enabled?) || no_reports?
```
[fixed] There are n+1 query warnings coming from Bullet on these type of lines, as `host` is lazily loaded.  You might possibly want to add an `:includes => [:host]` into build_global_status?  (Or optimise it even further, preventing another reload of Host).

domcleal commented on db/migrate/20150612135546_create_host_status.rb:
```diff
@@ -0,0 +1,32 @@
+    create_table :host_status do |t|
+      t.string :type
+      t.integer :status, :default => 0, :null => false
+      t.references :host, :null => false
```
I think this should be indexed, as this schema's going to cause a lot of loading of host_status records by host_id.

domcleal replied:
And it needs a foreign key, if it isn't generating that already.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -123,7 +123,7 @@ def destroy
 
       def status
-        render :json => { :status => @host.host_status }.to_json if @host
+        render :json => { :status => @host.get_status('HostStatus::ConfigurationStatus').to_label }.to_json if @host
```
Should we expand this to include more data about the detailed statuses, or create a new API?

ares replied:
I wanted to avoid bigger API changes, I suppose status should be global status and we should introduce new actions like configuration status but that would break V2 compatibility. So is this a good time for V3 and should we mix it into this PR?

domcleal replied:
Indeed, status here should probably be either global or Puppet specific.

Adding actions won't break v2 compatibility, unless you're going to significantly change/remove existing actions.  If there was an API to retrieve different types of status including global and this existing route was deprecated, then it'd probably stay returning the Puppet status.

domcleal commented on app/helpers/hosts_helper.rb:
```diff
@@ -72,44 +72,45 @@ def last_report_tooltip(record)
-      short = s_("OK|O")
+  def name_column(host)
+    style = host_global_status_icon_class_for_host(host)
+    tooltip = host.host_statuses.map { |status| "#{_(status.name)}: #{_(status.to_label)}" }.join(', ')
```
`host.host_statuses` may need some explicit ordering for consistency (here and elsewhere), since there's no default scope ordering.

domcleal commented on app/helpers/hosts_helper.rb:
```diff
@@ -270,6 +276,16 @@ def overview_fields(host)
 
+  def host_detailed_status_list(host, options = {})
+    content_tag(:ul, {:id => 'host_status_list'}.merge(options)) do
+      host.host_statuses.map do |status|
```
e.g. ordering here too

domcleal commented on app/models/host/managed.rb:
```diff
@@ -806,6 +815,33 @@ def setup_clone
+  def get_status(type)
+    status = host_statuses.find_by_type(type)
+    if status.nil?
+      status_class = type.constantize
```
I find this string-based API slightly strange, would it be more natural to pass the class/constant itself?

domcleal commented on app/helpers/hosts_helper.rb:
```diff
@@ -247,7 +248,12 @@ def show_templates
     fields = [
+      [_("Status"), content_tag(:i, '&nbsp'.html_safe, :class => host_global_status_icon_class(global_status.status)) +
+                    content_tag(:span, _(global_status.to_label), :class => host_global_status_class(global_status.status), :id => 'global_status_handler') +
+                    host_detailed_status_list(host, :class => 'hidden')
```
Should this be expanded by default?  I think unless we have hundreds of status types that showing them would be beneficial.

domcleal commented on app/helpers/hosts_helper.rb:
```diff
@@ -270,6 +276,16 @@ def overview_fields(host)
+      host.host_statuses.map do |status|
+        content_tag(:li,
+                    content_tag(:i, '&nbsp'.html_safe, :class => host_global_status_icon_class(status.to_global)) +
+                      content_tag(:span, status.to_label, :class => host_global_status_class(status.to_global), :title => _(status.name)))
```
`_(status.to_label)`

domcleal replied:
I'm also not terribly keen on only having the name in a tooltip, and along with the comment about about whether it should be expanded by default, would prefer something like a table:

Host status: OK
Build: Installed
Configuration: Something

domcleal commented:
1) n+1 queries: I commented about one of these (peculiarly, about HostStatus::Status accessing Host::Managed again), but the other is on HostsController or similar where an `:includes => [:host_statuses]` is required to preload the many statuses per host.  Perhaps just add this to included_associations in AppController, and also for the API controllers?

2) If a status changes from `relevant?` returning true to false, I don't see how it would be removed, e.g. a host becoming unmanaged.

3) Will status_registry cause any problems with class reloading, if it stores a set of classes?  It might need to store strings instead, IIRC?

4) HostStatus itself might get reloaded by the autoloader - seems that it might work today in core because it has a chain of require_dependency calls that would load the known classes, but perhaps not for plugins.  Some classes we keep outside of the autoloader, such as the menu loader and plugin registrations.

5) Should we put a method into Foreman::Plugin for access to registering a HostStatus type?

domcleal commented on app/models/host_status/build_status.rb:
```diff
@@ -0,0 +1,39 @@
+    def to_label
+      case to_status
+        when PENDING
+          N_("Pending Installation")
```
Lower case "i".

domcleal commented:
6) Will the build status refresh correctly?  The only event that I see a refresh event on is the report import, but it should probably hook into any change in the build attribute too.

ares commented:
> 1) n+1 queries: I commented about one of these (peculiarly, about HostStatus::Status accessing Host::Managed again), but the other is on HostsController or similar where an :includes => [:host_statuses] is required to preload the many statuses per host. Perhaps just add this to included_associations in AppController, and also for the API controllers?

fixed by adding :inverse_of and proper includes

> 2) If a status changes from relevant? returning true to false, I don't see how it would be removed, e.g. a host becoming unmanaged.

I think we shouldn't remove it but we should display only `relevant?` statuses? If the status was added by plugin which was later removed, it's the same issue as with settings or any other STI model, user has to clean up DB right?

> 3) Will status_registry cause any problems with class reloading, if it stores a set of classes? It might need to store strings instead, IIRC?

I don't think so since we use require_dependency in app/models/host_status.rb and app/models/host_status/status.rb

> 4) HostStatus itself might get reloaded by the autoloader - seems that it might work today in core because it has a chain of require_dependency calls that would load the known classes, but perhaps not for plugins. Some classes we keep outside of the autoloader, such as the menu loader and plugin registrations.

Plugins statuses should be under $plugin/app/ which is under autoloader control so I think it should work (not tested though).

> 5) Should we put a method into Foreman::Plugin for access to registering a HostStatus type?

Good idea, will add.

> 6) Will the build status refresh correctly? The only event that I see a refresh event on is the report import, but it should probably hook into any change in the build attribute too.

I'll add the refresh of build before_validation callback but I suppose we shouldn't even display this status in host detail if it's not `relevant?`

domcleal commented:
> > 2) If a status changes from relevant? returning true to false, I don't see how it would be removed, e.g. a host becoming unmanaged.
> 
> I think we shouldn't remove it but we should display only relevant? statuses?

Sure, yep.  Will the global status be calculated without it?  It looks like it would currently calculate it based on all host_statuses, without checking whether `relevant?` is still true.

> If the status was added by plugin which was later removed, it's the same issue as with settings or any other STI model, user has to clean up DB right?

Indeed, I think so.  We won't even be able to load the class to call `relevant?`.

> 6) Will the build status refresh correctly? The only event that I see a refresh event on is the report import, but it should probably hook into any change in the build attribute too.
> 
> I'll add the refresh of build before_validation callback but I suppose we shouldn't even display this status in host detail if it's not relevant?

Sure, but it will be relevant whether it's in build mode or not (I was thinking about a typical unattended host).

ares commented:
> > > 2) If a status changes from relevant? returning true to false, I don't see how it would be removed, e.g. a host becoming unmanaged.
> > 
> > I think we shouldn't remove it but we should display only relevant? statuses?
> 
> Sure, yep.  Will the global status be calculated without it?  It looks like it would currently calculate it based on all host_statuses, without checking whether `relevant?` is still true.

Right, fixed.

> > > 6) Will the build status refresh correctly? The only event that I see a refresh event on is the report import, but it should probably hook into any change in the build attribute too.
> > 
> > I'll add the refresh of build before_validation callback but I suppose we shouldn't even display this status in host detail if it's not relevant?
> 
> Sure, but it will be relevant whether it's in build mode or not (I was thinking about a typical unattended host).

Not sure I follow, we can update the status even for unmanaged host so if it's converted to managed the status will be there (I suppose marked as built)

domcleal commented:
> Not sure I follow, we can update the status even for unmanaged host so if it's converted to managed the status will be there (I suppose marked as built)

I'm only talking about unattended, managed hosts that are being provisioned.  When the build state changes, the global status needs updating.  I didn't see a mechanism for that.  Perhaps you've fixed that via the before_validation callback you mentioned.

I don't understand the relevant of your comments about unmanaged or !relevant? status hosts I'm afraid, I'm not talking about anything other than the regular provisioning case.

ares commented:
Rebased, fixes to review are in last commit

domcleal commented:
Tests are red.

ares commented:
Last pending item - add plugin API to register new status.

ares commented:
plugin API added and tested from remote execution plugin, tests should be green

EDIT: for inspiration how to add custom status from plugin, see https://github.com/ares/foreman_remote_execution/commit/a5a7433722ada98fd79ecdd425395b9fa22f37c4

domcleal commented on app/controllers/api/v1/hosts_controller.rb:
```diff
@@ -125,11 +125,45 @@ def destroy
+        render :json => { :status => @host.get_status(HostStatus::ConfigurationStatus).to_label }.to_json if @host
+      end
+
+      api :GET, "/hosts/:id/status/:type", "Get specific status of host"
```
I don't think there's any need to add new routes to v1, just make it available in v2 only.

ares commented:
Fixed, fingers crossed to get green tests again.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -123,7 +123,32 @@ def destroy
+      # TRANSLATORS: API documentation - do not translate
+      description 'Returns string representing a host status of a given type'
+      def get_status
+        case params[:type]
```
How will plugins get their status types recognised here?  Shouldn't it be looked up in the registry?

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -123,7 +123,32 @@ def destroy
+        render :json => { :status => @host.get_status(HostStatus::ConfigurationStatus).to_label }.to_json if @host
+      end
+
+      api :GET, "/hosts/:id/status/:type", "Get status of host"
```
strings should all be extracted

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -123,7 +123,32 @@ def destroy
+* configuration
+* build
+      eos
+      # TRANSLATORS: API documentation - do not translate
```
Should be translated in v2

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -123,7 +123,32 @@ def destroy
+          else
+            raise ::Foreman::Exception.new(N_("Invalid type for host status: %s"), params[:type])
+        end
+        render :json => { :status => status.to_label }.to_json
```
Consider using rabl, it might be easier to extend for plugins.

domcleal replied:
In particular I think it would be useful to return both the integer (as host#show does) and the label, since it's a mix right now.  The value can also be translated.

domcleal commented:
Please wait until I've finished reviewing before updating, I'll always change the label.  It's harder when it's changing!

domcleal commented on app/models/host/managed.rb:
```diff
@@ -848,6 +859,41 @@ def setup_clone
+  def configuration_status
+    get_status(HostStatus::ConfigurationStatus).to_status
+  end
+  alias_method :puppet_status, :configuration_status
```
Consider deprecating puppet_status.

domcleal commented on app/views/api/v2/hosts/main.json.rabl:
```diff
@@ -9,7 +9,12 @@ attributes :ip, :environment_id, :environment_name, :last_report, :mac, :realm_i
-           :last_compile, :puppet_status
+           :last_compile, :global_status, :puppet_status
+
+if @host
```
When is it false?  Perhaps use rabl's conditionals instead: https://github.com/nesquena/rabl#conditions

ares replied:
When main is loaded in index action

domcleal replied:
Should this be in the show template then?

ares commented:
> Please wait until I've finished reviewing before updating, I'll always change the label. It's harder when it's changing!

Sorry I didn't expect many comments in repeated review...

domcleal commented on db/migrate/20150612135546_create_host_status.rb:
```diff
@@ -0,0 +1,36 @@
+    add_foreign_key "host_status", "hosts", :name => "host_status_hosts_host_id_fk", :column => 'host_id'
+    add_column :hosts, :global_status, :integer, :default => 0, :null => false
+
+    Host.all.each do |host|
```
Host.all will iterate only over Host::Managed I think, but this feature seems to be on Host::Base.  Should it loop over the latter?

ares replied:
`refresh_statuses` is defined in `Host::Managed`

domcleal replied:
Should the has_many move to Host::Managed then?

ares replied:
probably yes, will move

lzap replied:
Is there a reason why the status methods are not in Host::Base? This utterly broke discovery now.

domcleal replied:
Is there a reason they should be?  Could you please file bugs in Redmine so we're aware of precisely what broke?

lzap replied:
> Is there a reason they should be?  Could you please file bugs in Redmine so we're aware of precisely what broke?

For the record, discussion moved here:
http://projects.theforeman.org/issues/11708

## 

Later,
 Lukas #lzap Zapletal

domcleal commented on app/models/host/managed.rb:
```diff
@@ -33,6 +32,8 @@ def self.complete_for(query, opts = {})
   define_model_callbacks :build, :only => :after
   define_model_callbacks :provision, :only => :before
 
+  before_validation :refresh_build_status, :if => :build_changed?
```
I think this is missing a unit test.

domcleal replied:
Ah no, I think it's covered by the waiting_to_build? test.

ares replied:
should be covered here https://github.com/theforeman/foreman/pull/2580/files#diff-88c8d0302188366e58cfd358acf05756R437

domcleal commented on test/unit/host_status/configuration_status_test.rb:
```diff
@@ -0,0 +1,67 @@
+  end
+
+  test '#out_of_sync? is true if reported_at is set and is too long ago' do
+    assert_present @status.reported_at
```
Nitpick, Rails 4 removes assert_present so use `assert foo.present?` instead.

domcleal commented on test/unit/host_status/configuration_status_test.rb:
```diff
@@ -0,0 +1,67 @@
+    Setting[:outofsync_interval] = original
+  end
+
+  test '#refresh! update date to reported_at of last report' do
```
Is this desc correct?  It looks like it belongs to the test below.

domcleal commented on test/unit/host_status_test.rb:
```diff
@@ -0,0 +1,10 @@
+    status = OpenStruct.new
+    HostStatus.status_registry.add(status)
+    assert_includes HostStatus.status_registry, status
+    HostStatus.status_registry.delete(status)
```
Perhaps refute that the status is registered.

ares replied:
added (I suppose you meant just to detect the issue early, not really a test)

domcleal commented on app/models/host_status/global.rb:
```diff
@@ -0,0 +1,42 @@
+
+    def self.build(statuses)
+      max_status = statuses.select { |s| s.relevant? }.map do |status|
+        status.to_global
```
`.map(&:to_global)` ?

domcleal commented on app/assets/javascripts/application.js:
```diff
@@ -127,6 +127,10 @@ function onContentLoad(){
     $(this).find('.form_template').remove()
   })
+
+  $('#global_status_handler').on('click', function(){
```
Unused now that they're in the table?

domcleal commented on app/assets/stylesheets/application.scss:
```diff
@@ -546,3 +546,7 @@ table {
   margin-right: 10px;
 }
+
+#global_status_handler {
```
Unused?

domcleal commented on app/helpers/hosts_helper.rb:
```diff
@@ -247,7 +248,14 @@ def show_templates
+    global_status = host.build_global_status
     fields = [
+      [_("Status"), content_tag(:i, '&nbsp'.html_safe, :class => host_global_status_icon_class(global_status.status)) +
+                    content_tag(:span, _(global_status.to_label), :class => host_global_status_class(global_status.status), :id => 'global_status_handler')
```
ID unused if the JS/CSS are unused?

domcleal commented on app/helpers/hosts_helper.rb:
```diff
@@ -270,6 +278,17 @@ def overview_fields(host)
+      [
+        _(status.name),
+        content_tag(:i, '&nbsp'.html_safe, :class => host_global_status_icon_class(status.to_global)) +
+          content_tag(:span, _(status.to_label), :class => host_global_status_class(status.to_global), :title => _(status.name))
```
I don't know if title is any use here, since it's in the left hand column.

domcleal commented on app/helpers/hosts_helper.rb:
```diff
@@ -270,6 +278,17 @@ def overview_fields(host)
+      next unless status.relevant?
+      [
+        _(status.name),
+        content_tag(:i, '&nbsp'.html_safe, :class => host_global_status_icon_class(status.to_global)) +
```
I think this is adding an unnatural amount of spacing between the icon and status text?

![screenshot from 2015-08-20 12-23-43](https://cloud.githubusercontent.com/assets/482089/9382195/5b18add4-4736-11e5-956c-16f78c8e7bbf.png)

domcleal commented on app/helpers/hosts_helper.rb:
```diff
@@ -72,44 +72,45 @@ def last_report_tooltip(record)
-      short = s_("OK|O")
+  def name_column(host)
+    style = host_global_status_icon_class_for_host(host)
+    tooltip = host.host_statuses.sort_by(&:type).map { |status| "#{_(status.name)}: #{_(status.to_label)}" }.join(', ')
```
I think this will also show non-relevant statuses.

domcleal commented on app/views/api/v2/hosts/main.json.rabl:
```diff
@@ -9,7 +9,12 @@ attributes :ip, :environment_id, :environment_name, :last_report, :mac, :realm_i
+           :last_compile, :global_status, :puppet_status
+
+if @host
+  attributes :configuration_status if @host.get_status(HostStatus::ConfigurationStatus).relevant?
```
I think including the translated text would be useful too, as it can be shown in things like `hammer host info`, plus it'd be more consistent with the new status route.

domcleal commented:
Is it deliberate that BuildStatus doesn't override to_global?  It means we lose any immediate visual indication on the host list that a host is in the build/pending state, you can only tell via the tooltip.  It returns an OK global status currently.  I guess Global doesn't really have the means to express "pending" and more subtle statuses.

Thanks, looking good.

ares commented:
> Is it deliberate that BuildStatus doesn't override to_global? It means we lose any immediate visual indication on the host list that a host is in the build/pending state, you can only tell via the tooltip. It returns an OK global status currently. I guess Global doesn't really have the means to express "pending" and more subtle statuses.

Yes, I think that pending build does not mean anything you should worry about, so we don't even map it to global warning. If you're interested in it, you can find such hosts easily. Hopefully we'll have customizable columns soon so user could add build status column if it's important.

ares commented:
Everything should be addressed, if you could give it another round even before tests I'd appreciate. I'd like to leave it ready so it does not hang on me while I'm AFK.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -109,7 +109,7 @@ def destroy
       end
 
-      api :GET, "/hosts/:id/status", N_("Get status of host")
+      api :GET, "/hosts/:id/status", N_("Get specific status of host")
```
configuration status?

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -123,7 +123,25 @@ def destroy
+
+      api :GET, "/hosts/:id/status/:type", N_("Get status of host")
+      param :id, :identifier_dottable, :required => true
+      param :type, [ HostStatus::Global ] + HostStatus.status_registry.to_a.map { |s| s.humanized_name }, :required => true, :desc => N_('
```
Use heredoc here, I think the string extraction things understand it better.

domcleal commented on app/models/host_status.rb:
```diff
@@ -0,0 +1,11 @@
+    @status_registry ||= Set.new
+  end
+
+  def self.find_status_by_humanized_name(name)
```
Missing test coverage I think

ares replied:
added here https://github.com/ares/foreman/commit/366fca29f03d921564e7dde83be9508e6b08548d#diff-2e021e3b74348acab0c3703cf11e34e8R18

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -123,7 +123,25 @@ def destroy
+* configuration
+* build')
+      description N_('Returns string representing a host status of a given type')
+      def get_status
```
Missing functional test coverage.

ares replied:
https://github.com/theforeman/foreman/pull/2580/files#diff-a7b88f691f88ff3665fe4f0f563fb4d6R183 - or do you want me to cover more status types?

domcleal commented on app/views/api/v2/hosts/main.json.rabl:
```diff
@@ -9,7 +9,10 @@ attributes :ip, :environment_id, :environment_name, :last_report, :mac, :realm_i
            :compute_profile_id, :compute_profile_name, :capabilities, :provision_method,
            :puppet_proxy_id, :certname, :image_id, :image_name, :created_at, :updated_at,
-           :last_compile, :puppet_status
+           :last_compile, :global_status, :global_status_label, :puppet_status
```
You might need to use a rabl alias to display puppet_status so it doesn't hit the method deprecation.

domcleal commented on app/views/api/v2/hosts/main.json.rabl:
```diff
@@ -9,7 +9,10 @@ attributes :ip, :environment_id, :environment_name, :last_report, :mac, :realm_i
-           :last_compile, :puppet_status
+           :last_compile, :global_status, :global_status_label, :puppet_status
+
+attributes :configuration_status, :configuration_status_label, :if => @host && @host.get_status(HostStatus::ConfigurationStatus).relevant?
```
Commented on this last time, but I think these should move into the show rabl if they're not meant to be displayed on the index.

domcleal replied:
Should it also just loop around all host_statuses (where relevant) and add attributes for them all?  Then it'll automatically be extended for plugins adding statuses without any extra changes required.

domcleal commented on app/controllers/application_controller.rb:
```diff
@@ -338,7 +338,7 @@ def check_empty_taxonomy
   # We do not include most associations unless we are processing a html page
   def included_associations(include = [])
-    include += [:hostgroup, :compute_resource, :operatingsystem, :environment, :model ]
+    include += [:hostgroup, :compute_resource, :operatingsystem, :environment, :model, :reports, :host_statuses ]
```
Including :reports here is going to be a really expensive call:

```
2015-08-21T12:34:00 [sql] [D]   Report Load (9.7ms)  SELECT "reports".* FROM "reports" WHERE "reports"."host_id" IN (110, 109, 170, 82, 168, 118)
```

If a host has many reports, it's going to pull back a potentially huge number of records.

domcleal commented on app/models/host/base.rb:
```diff
@@ -23,6 +23,7 @@ class Base < ActiveRecord::Base
             :conditions => { :provision => true }
     has_one :domain, :through => :primary_interface
     has_one :subnet, :through => :primary_interface
+    has_many :host_statuses, :class_name => 'HostStatus::Status', :foreign_key => 'host_id', :inverse_of => :host
```
I think this is missing a `:dependent => :destroy`, as status entries aren't being deleted on host deletion.

domcleal commented:
Tiny one this, but the vertical alignment of the status icons on both the index and host show pages have changed.  The text used to be centred vertically in the status icon, but is now bottom-aligned.

![screenshot from 2015-08-21 12-33-15](https://cloud.githubusercontent.com/assets/482089/9407486/88e6f164-4801-11e5-8bc3-899edf1e19f1.png)

![screenshot from 2015-08-21 12-35-01](https://cloud.githubusercontent.com/assets/482089/9407487/8ad2f9c8-4801-11e5-9049-fec751018264.png)

ares commented:
> The text used to be centred vertically in the status icon, but is now bottom-aligned.

If someone tells me how to do that with bootstrap glyphicon I'm happy to do that. I tried the same styling but it does not work, I suppose it's because it's not an HTML text that we try to shift down. I could shrink the icon though like this
![icons](https://cloud.githubusercontent.com/assets/109773/9542819/5eadee84-4d74-11e5-8f9e-62fe5b377b07.png)

ares commented:
@domcleal thanks for many good points, I think I addressed all except the vertical align thing, see if the attached screenshot would be better (it's not included).

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -123,7 +123,27 @@ def destroy
+      api :GET, "/hosts/:id/status/:type", N_("Get status of host")
+      param :id, :identifier_dottable, :required => true
+      param :type, [ HostStatus::Global ] + HostStatus.status_registry.to_a.map { |s| s.humanized_name }, :required => true, :desc => N_(<<-eos
+status type, can be one of configuration
```
Is the word "configuration" meant to be here?

ares replied:
definitely not :-)

domcleal commented:
> If someone tells me how to do that with bootstrap glyphicon I'm happy to do that. I

Perhaps by changing the `top` attribute?  It seems it's set to 1px in bootstrap_and_overrides right now, but increasing it to 3px looks a bit more normal to me - probably just because the text is typically lower case on the host list.

ares commented:
> Perhaps by changing the top attribute?

works! will push when you switch the label

domcleal commented:
The host#index page still has a SQL query that could cause poor performance.  It seems to come from last_report_object, which ought to be optimised, but isn't.

When I load it on this branch, I see these queries:

<pre>
2015-08-28T13:01:25 [sql] [D]   Report Load (4.8ms)  SELECT "reports".* FROM "reports" WHERE "reports"."host_id" IN (110, 109, 170, 82, 168, 118) ORDER BY reports.id DESC
2015-08-28T13:01:25 [sql] [D]   HostStatus::Status Load (0.3ms)  SELECT "host_status".* FROM "host_status" WHERE "host_status"."host_id" IN (110, 109, 170, 82, 168, 118)
2015-08-28T13:01:25 [sql] [D]    (0.3ms)  SELECT MAX("reports"."id") AS maximum_id, host_id AS host_id FROM "reports" WHERE "reports"."host_id" IN (110, 109, 170, 82, 168, 118) GROUP BY host_id
</pre>

The first comes from including last_report_object, as if I remove it from the list in application controller, it instead does this:

<pre>
2015-08-28T13:06:18 [sql] [D]   Report Load (0.1ms)  SELECT "reports".* FROM "reports" WHERE "reports"."host_id" = 170 ORDER BY reports.id DESC LIMIT 1
2015-08-28T13:06:18 [sql] [D]   Report Load (0.2ms)  SELECT "reports".* FROM "reports" WHERE "reports"."host_id" = 82 ORDER BY reports.id DESC LIMIT 1
</pre>

which also triggers Bullet to warn about the N+1 query.

The third well-optimised query in the first paste is actually from the controller over here: https://github.com/theforeman/foreman/blob/1.9.0/app/controllers/hosts_controller.rb#L43.  This is used for the "last report" column.  Can we reuse that somehow?

ares commented:
@domcleal fixed new issues, reusing of @last_reports does not look nice especially because it makes global status coupled with configuration status but I couldn't think of better way, ideas welcome as always

domcleal commented on app/models/host_status/global.rb:
```diff
@@ -0,0 +1,44 @@
+    attr_accessor :status
+
+    def self.build(statuses, options = {})
+      if options.has_key?(:last_reports) && (configuration = statuses.find { |s| s.type == 'HostStatus::ConfigurationStatus' })
```
`.is_a?` rather than string comparison?

domcleal commented on app/models/host_status/global.rb:
```diff
@@ -0,0 +1,44 @@
+    def self.build(statuses, options = {})
+      if options.has_key?(:last_reports) && (configuration = statuses.find { |s| s.type == 'HostStatus::ConfigurationStatus' })
+        cached_report = options[:last_reports].find { |r| r.host_id == configuration.host_id }
+        configuration.last_report = cached_report
```
Could we just pass all options into a status, e.g. into to_global as an argument or another initialisation sort of method?  This may come up again if new statuses are added that require looking at other DB records.

domcleal commented:
> reusing of @last_reports does not look nice especially because it makes global status coupled with configuration status but I couldn't think of better way, ideas welcome as always

Yeah, it's rather awful, but vital I think.  Your implementation of explicitly passing it in is reasonable.  I think though it's going to need duplicating to the API host controllers, as they're showing the same sort of problems with fetching all indexes and host_statuses.

ares commented:
>  I think though it's going to need duplicating to the API host controllers, as they're showing the same sort of problems with fetching all indexes and host_statuses.

I didn't see the same issue, just the regular n+1 (which is there for many other associations). I improved what I could, with rabl it gets "a bit" dirty.

ares commented:
Rebased the branch, since last time, last 3 commits were added. Please let me know whether I can squash it since rebasing becomes very unpleasant.

domcleal commented:
Sure, please squash and rebase, looks fine.

ares committed:
Fixes #10782 - global host status

Introduce new global host status that is composed of host substatuses.
Each substatus defines a mapping to the global one which can result in
three values
* OK
* WARN
* ERROR

Plugins can add their own substatuses. These are automatically
propagated also to API.

Thanks to Tomas Strachota who wrote the original code.

ares commented:
Squashed and rebased

domcleal commented on app/models/concerns/hostext/search.rb:
```diff
@@ -15,13 +16,16 @@ module Search
-      scoped_search :on => :puppet_status, :offset => Report::METRIC.index("pending"),         :word_size => Report::BIT_NUM, :rename => :'status.pending'
+
+      scoped_search :in => :host_statuses, :on => :status, :offset => 0, :word_size => Report::BIT_NUM*4, :rename => :'status.interesting', :complete_value => {:true => true, :false => false}
+      scoped_search_status "applied",         :in => :host_statuses, :on => :status, :rename => :'status.applied'
```
Should these be limited to ConfigurationStatus?  I think they're matching any status right now, including stuff like the build status.

domcleal commented:
There's a stray `a.txt` file in your squashed commit.

ares commented:
Removed a.txt (sorry I needed to inject some commit during squashing and forgot to remove it) and introduced a new relation to configuration_status which is now used in scoped search queries.

domcleal commented:
The tests are red, a page is failing with internal server errors.

ares commented:
@domcleal I'm working on that, I hoped you won't notice before I fix it (rebase issue after stabby lamda change...)

ares commented:
:green_apple:

domcleal commented on app/models/concerns/configuration_status_scoped_search.rb:
```diff
@@ -0,0 +1,10 @@
+
+  module ClassMethods
+    def scoped_search_status(status, options)
+      options.merge({ :offset => Report::METRIC.index(status.to_s), :word_size => Report::BIT_NUM })
```
`merge!` else it doesn't work at all.

domcleal commented:
Thanks @ares and @tstrachota, merged as e54016d.

Could you add a section to http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin about how to create your own substatus?  This could be useful for a variety of plugins.

ares commented:
For the reference, the section was added [here](http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin/diff?utf8=%E2%9C%93&version=72&version_from=71&commit=View+differences)
